ui/ducks/send/send.js

```diff
@@ -0,0 +1,1442 @@
+        state.gas.gasPrice = action.payload.gasPrice;
+        state.gas.gasTotal = action.payload.gasTotal;
+        if (state.stage !== SEND_STAGES.UNINITIALIZED) {
+          slice.caseReducers.validateRecipientUserInput(state, {
+            payload: {
+              chainId: action.payload.chainId,
+              tokens: action.payload.tokens,
+            },
+          });
+        }
+        state.stage =
+          state.stage === SEND_STAGES.UNINITIALIZED
+            ? SEND_STAGES.ADD_RECIPIENT
+            : state.stage;
+        slice.caseReducers.validateAmountField(state);
+        slice.caseReducers.validateGasField(state);
+        slice.caseReducers.validateSendState(state);
+      })
+      .addCase(computeEstimatedGasLimit.pending, (state) => {
+        // When we begin to fetch gasLimit the slice should be marked invalid
+        state.status = SEND_STATUSES.INVALID;
+      })
+      .addCase(computeEstimatedGasLimit.fulfilled, (state, action) => {
+        // When we receive a new gasLimit from the computeEstimatedGasLimit
+        // thunk we need to update our gasLimit in the slice. We call into the
+        // caseReducer updateGasLimit to tap into the appropriate follow up
+        // checks and gasTotal calculation.
+        if (action.payload?.gasLimit) {
+          slice.caseReducers.updateGasLimit(state, {
+            payload: action.payload.gasLimit,
+          });
+        }
+      })
+      .addCase(SET_BASIC_GAS_ESTIMATE_DATA, (state, action) => {
+        // When we receive a new gasPrice via the gas duck we need to update
+        // the gasPrice in our slice. We call into the caseReducer
+        // updateGasPrice to also tap into the appropriate follow up checks
+        // and gasTotal calculation.
+        slice.caseReducers.updateGasPrice(state, {
+          payload: getGasPriceInHexWei(action.value.average ?? '0x0'),
+        });
+      })
+      .addCase(BASIC_GAS_ESTIMATE_STATUS, (state, action) => {
+        // When we fetch gas prices we should temporarily set the form invalid
+        // Once the price updates we get that value in the
+        // SET_BASIC_GAS_ESTIMATE_DATA extraReducer above. Finally as long as
+        // the state is 'READY' we will revalidate the form.
+        switch (action.value) {
+          case BASIC_ESTIMATE_STATES.FAILED:
+            slice.caseReducers.useInlineGasFields(state);
+            state.status = SEND_STATUSES.INVALID;
+            break;
+          case BASIC_ESTIMATE_STATES.LOADING:
+            state.status = SEND_STATUSES.INVALID;
+            break;
+          case BASIC_ESTIMATE_STATES.READY:
+          default:
+            slice.caseReducers.validateSendState(state);
+        }
+      })
+      .addCase(SET_ESTIMATE_SOURCE, (state, action) => {
+        if (action.value === GAS_SOURCE.ETHGASPRICE) {
+          slice.caseReducers.useInlineGasFields(state);
+        } else {
+          slice.caseReducers.useBasicGasFields(state);
```

Comment:
This seems like it might cause the basic gas fields to be shown when they shouldn't be. Like if the user is on a custom network, or has advanced gas enabled, or has already chosen a custom gas price.

Reply:
there's only two sources possible here, ETHGASPRICE and METASWAPS

we only hit the former for custom networks, making it impossible for this reducer to flip to basic on custom networks. 

the getGasInputMode selector will return 'INLINE' if advanced gas is enabled (effectively ignoring the value from state, not ideal but i was struggling with how to incorporate this setting into the extraReducers here)

I'm going to add a check for the mode to ensure that it won't do anything if a custom gas price is already set.

Reply:
https://github.com/MetaMask/metamask-extension/pull/10965/commits/57de0f70344269c482c764623538166a17547845